metals/src/main/scala/scala/meta/internal/metals/Compilations.scala

```diff
@@ -107,24 +108,28 @@ final class Compilations(
     compileBatch(targets).ignoreValue
   }
 
-  def compileFile(path: AbsolutePath): Future[b.CompileResult] = {
-    def empty = new b.CompileResult(b.StatusCode.CANCELLED)
-    for {
-      targetOpt <- expand(path)
-      result <- targetOpt match {
-        case None => Future.successful(empty)
-        case Some(target) =>
-          compileBatch(target)
-            .map(res => res.getOrElse(target, empty))
-      }
-      _ <- compileWorksheets(Seq(path))
-    } yield result
+  def compileFile(path: PathWithContent): Future[Option[b.CompileResult]] = {
+    if (fileSignatures.didSavedContentChanged(path)) {
+      def empty = new b.CompileResult(b.StatusCode.CANCELLED)
+      for {
+        targetOpt <- expand(path.path)
+        result <- targetOpt match {
+          case None => Future.successful(empty)
+          case Some(target) =>
+            compileBatch(target)
+              .map(res => res.getOrElse(target, empty))
+        }
+        _ <- compileWorksheets(Seq(path.path))
+      } yield Some(result)
+    } else Future.successful(None)
   }
 
   def compileFiles(
-      paths: Seq[AbsolutePath],
+      pathsWithContent: Seq[PathWithContent],
       focusedDocumentBuildTarget: Option[BuildTargetIdentifier],
   ): Future[Unit] = {
+    val paths =
```

Comment:
Maybe we could just find first that changed in a given target?

Reply:
Yes, though I doubt it ever too many files at once. Only file watcher passes more than one file.